[1단계 - 로또 구현] 새로이(강철원) 미션 제출합니다.

Ryan-Dia opened the pull request:
## 😰 아쉬운 점

1. 통합 테스트를 고려하지 않고 구현한 점

- 전략패턴을 사용하려고 하였으나 시간관계상 테스트가 불가능하지만 바로 구현할 수 있는 방법으로 선택했습니다.
  
```java
    public void generate(Money money) {
        // TODO 통합 테스트 불가능
        IntStream.range(0, money.getLottoTicketCount())
                .forEach(index -> {
                    lottoGroup.add(Lotto.create(RandomNumberUtils.generateRandomNumbers()
                            .stream()
                            .map(LottoNumber::new)
                            .toList()));
                });
    }
```

2. 토론에 집중하느라 기능구현 밖에 하지 못해서 테스트를 추가하지 못한 점
3. 에러메시지를 상수화해서 관리하지 못한 점

## 🚀궁금한 점

- 변수명이 하나 일때 클래스명과 동일해도 되는지 궁금합니다.
```java
public class LottoGroup {
    private final List<Lotto> lottoGroup = new ArrayList<>();
```
- 매개변수 타입에 대한 논쟁이 있어서 리뷰이님은 어떻게 생각하시는지 궁금합니다.
```java
 /**
     * // NOTICE 매개변수 사용에대한 의견 대립 List<LottoNumbers> - 필요한 것만 받을 수 있다. 즉 객체 전체를 넘겨주지 않아도 되기 때문에 Lotto 객체를 은닉화할 수 있다.
     * Lotto lotto - 강력한 타입 제한, Lotto의 객체를 통해 로또의 List<LottoNumbers>의 제약을 걸 수 있음.
     */
    public long getMatchCount(Lotto lotto) {
        return lotto.getLottoNumbers().stream().filter(winnerNumbers::contains).count();
    }
```

- 아래의 `toLottoNumberList` 메서드를 일단 `Lotto` 객체에 위치시켰으나  `LottoNumber`에 있는게 더 좋은지 아니면 다른 곳에 있으면 좋은지 궁금합니다. (다른 더 좋은 방법이 있는지도 궁금합니다)
```java
private List<LottoNumber> toLottoNumberList(String input) {
        if (input == null || input.isBlank() || input.endsWith(", ")) {
            throw new IllegalArgumentException("잘못된 입력입니다. 이와 같은 형태로 작성해주세요.(ex. 1, 2, 3, 4, 5, 6)");
        }
        return Arrays.stream(input.split(", "))
                .map(LottoNumber::new)
                .toList();
    }
```

## 📌 프롤로그 학습로그
https://prolog.techcourse.co.kr/studylogs/3866

Ryan-Dia committed:
docs: 기능 목록 추가

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
feat : 로또 도메인 구현

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
feat: 재시도 유틸리티 추가

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
feat: 수익 관련 객체 추가

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
feat: 프로젝트에 사용되는 유틸리티 구현

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
feat: 로또 당첨 관련 도메인 구현

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
chore: lotto 디렉토리로 파일 이동

Ryan-Dia committed:
feat: toString 추가

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: 에러 타입 변경

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: 수익률 변수 초기화 수정

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
feat: view 추가

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
feat: 줄바꿈 상수화

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
feat: 로또 컨트롤러 추가

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
merge change

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: 인수로 객체 대신 DTO 전달하도록 변경

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: 인수에 객체대신 DTO 전달

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
style: 사용하지 않는 import 제거

Ryan-Dia committed:
refactor: 사용하지 않는 메서드 제거

Ryan-Dia committed:
refactor: 매직넘버 상수화

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: 매직넘버 상수화

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: 매직넘버 상수화

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: 에러 메시지 수정

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: 매직넘버 상수화

Ryan-Dia committed:
refactor: 유틸 클래스 방어적으로 변경

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

Ryan-Dia committed:
refactor: toLottoNumberList 메서드를 Lotto 클래스로 이동

Co-authored-by: kiwoook <tizmfns1218@Naver.com>

dusdn1702 commented on src/main/java/lotto/common/Constants.java:
```diff
@@ -0,0 +1,12 @@
+    public static final int LOTTO_NUM_SIZE = 6;
+    public static final int MAX_LOTTO_NUMBER = 45;
+    public static final int MIN_LOTTO_NUMBER = 1;
+    public static final String ENTER = System.lineSeparator();
```
라인 구분자에 엔터라는 이름이 적합할지 고민해보면 좋을 것 같아요!

Ryan-Dia replied:
네 맞습니다. 너무 고민을 안 했던 것 같습니다.
ENTER보다는 LINE_SEPARATOR더 좋을 것 같아 수정했습니다. 
감사합니다.

dusdn1702 commented on src/main/java/lotto/common/Constants.java:
```diff
@@ -0,0 +1,12 @@
+    public static final int MIN_LOTTO_NUMBER = 1;
+    public static final String ENTER = System.lineSeparator();
+
+    private Constants() {
+    }
```
이렇게 선언해주신 이유가 있을까요?

Ryan-Dia replied:
`Constants` 클래스는 정적 변수만 사용되기 때문에  인스턴스가 필요하지 않습니다. 
따라서 불필요한 인스턴스화를 방지하기 위해 `private` 생성자를 추가하였습니다.  

더불어 `private` 생성자가 불필요한 상속도 막아주기 때문에 1석2조의 효과를 얻을 수 있기에 사용했습니다.


이와관련해서는 어제/오늘해서 블로그 글을 작성했습니다.
[private 생성자를 왜 만드시나요?](https://html-jc.tistory.com/754)

dusdn1702 replied:
간단한 규칙으로 적용하는 코드 그 이상의 답변을 하지 못해서 파생되는 개념까지 공부하고, 정리까지 하셨다니 정말 멋지네요! 😲 
저도 많이 배웠습니다 ㅎㅎ 👍

dusdn1702 commented on src/main/java/lotto/common/Constants.java:
```diff
@@ -0,0 +1,12 @@
+package lotto.common;
+
+public final class Constants {
```
상수 분리에 불변 선언까지! 👍

dusdn1702 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,76 @@
+        return RecoveryUtils.executeWithRetry(InputView::readMoney, Money::new);
+    }
+
+    private void generateLotto(Money money) {
```
메서드명만 보면 Lotto라는 응답 객체가 기대될 것 같아요!

Ryan-Dia replied:
다시보니 정말 그렇다고 생각합니다. 
해서 메서드명을 수정하였습니다.

dusdn1702 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,61 @@
+        return lottoNumbers;
+    }
+
+    public static Lotto create(final List<LottoNumber> lottoNumbers) {
```
static으로 객체 생성해주신 이유가 있을까요?

Ryan-Dia replied:
처음의도는 정적팩터리 메서드를 사용하려고 사용했으나 현재 상태에서는 해당 장점이 없는데도 관성적으로 사용했던 것 같습니다.
그렇기에 삭제했습니다.

dusdn1702 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,48 @@
+
+
+    private boolean filter(int correctCount, boolean hasBonusNumber) {
+        if (correctCount == 5) {
```
이 숫자는 상수화 안하신 기준이 있으실까요?

Ryan-Dia replied:
저희가 놓쳤습니다.  
캐치해주셔서 감사합니다. 
해당부분 수정했습니다.

dusdn1702 commented on src/main/java/lotto/domain/WinnerLotto.java:
```diff
@@ -0,0 +1,40 @@
+     * Lotto lotto - 강력한 타입 제한, Lotto의 객체를 통해 로또의 List<LottoNumbers>의 제약을 걸 수 있음.
+     */
+    public long getMatchCount(Lotto lotto) {
+        return lotto.getLottoNumbers().stream().filter(winnerNumbers::contains).count();
```
한줄에 체이닝이 길어지면 가독성이 떨어지지 않을까요?

Ryan-Dia replied:
네 맞습니다.  그래서 해당부분을 가독성 있도록 수정했습니다.
더불어 `getLottoNumbers` 즉, getter없이 구현할 아이디어가 생각나서 함께 수정했습니다.

dusdn1702 commented on src/main/java/lotto/utils/NumberUtils.java:
```diff
@@ -0,0 +1,15 @@
+        try {
+            return Integer.parseInt(number);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("숫자만 가능합니다.");
```
에러 처리를 이렇게 해주신 이유가 궁금합니다~

Ryan-Dia replied:
`parseInt` 메서드에 문자 같은 것을 주입하면 내부적으로 NumberFormatException 예외를 던져줍니다. 
하지만 저희는 해당로직이 오직 사용자가 숫자가아닌 문자를 썼을 때 왜 잘못되었는지 표기하기 위함과 사용자 잘못이라는 의미를 남기기 위해 `IllegalArgumentException`을 사용했습니다.

dusdn1702 commented on src/main/java/lotto/utils/RecoveryUtils.java:
```diff
@@ -0,0 +1,24 @@
+                return processFunction.apply(inputSupplier.get());
+            } catch (IllegalArgumentException e) {
+                OutputView.printError(e);
+
```
요기 컨벤션이 잘 지켜지지 않은 것 같네요!

dusdn1702 commented on src/main/java/lotto/utils/RecoveryUtils.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public static <T, R> R executeWithRetry(Supplier<T> inputSupplier, Function<T, R> processFunction) {
+        while (true) {
```
계속 에러가 발생하면 성공할 때까지 에러가 쌓이는걸 의도하신걸까요? 🙄

Ryan-Dia replied:
의도한건아닙니다. 
"그냥 실패했을 때 재시도를 할 수 있게 만들자"만 고려했던 것 같습니다. 
사실 이게 서비스라면 저런식으로 무한반복하게 만들면 안 된다고 생각합니다. 
무한 반복이 가능하다면 서버에서 CPU와 로그를 과도하게 잡아먹을 수 있다고 생각합니다.  

따라서 최대 재시도횟수를 5회로 제한했습니다.

dusdn1702 replied:
사용자 측면에서 서비스를 구현할지, 관리자 측면에서 서비스를 구현할지 매번 고민되는 부분이죠! 👍

dusdn1702 reviewed (changes requested):
안녕하세요 새로이 
첫 리뷰를 맡게 된 샐리라고 합니다 😀

로또 미션 정말 잘 구현해주셨어요 고생 많으셨습니다 ~
리뷰하면서 궁금했던 점들 몇가지 남겨두었습니다 확인 부탁드릴게요 !

dusdn1702 commented:
남겨주신 질문도 답변드립니다 🙇🏻‍♀️

- 변수명이 하나 일때 클래스명과 동일해도 되는지 궁금합니다.
동일하면 안된다, 된다로 나눌 수는 없겠지만 저는 동일하게 두지 않을 것 같아요.
가장 큰 이유는 가독성 측면인데요, lottoGroup을 호출했을 때 변수를 호출했는지, 클래스를 호출했는지 알기 어렵지 않을까요? 
객체가 하는 역할과 변수가 하는 역할에 대해 생각해보면 다른 이름이 나올 수 있을 것 같아요 👍 

- 매개변수 타입에 대한 논쟁이 있어서 리뷰이님은 어떻게 생각하시는지 궁금합니다.
양쪽의 의견이 다 이해가 되는데요~!
저라면 이렇게 논쟁이 있다는 것은 코드 구조 개선을 통해 해결할 수 있는게 아닌지 다시 확인해볼 것 같아요.
WinnerLotto와 Lotto에 어떠한 연관이 있는지 왜 Lotto를 은닉화해야하는지 WinnerLotto는 어떤 역할을 가졌는지 한번 생각해볼까요?

- 아래의 toLottoNumberList 메서드를 일단 Lotto 객체에 위치시켰으나 LottoNumber에 있는게 더 좋은지 아니면 다른 곳에 있으면 좋은지 궁금합니다. (다른 더 좋은 방법이 있는지도 궁금합니다)
입력값을 객체로 변환해주는 메서드로 보이는데요~ 이게 Lotto의 역할이 맞을까요? 
입력값에 대한 유효성과 객체 생성을 분리해보는 건 어떨까요?

dusdn1702 commented on src/main/java/lotto/utils/RecoveryUtils.java:
```diff
@@ -0,0 +1,24 @@
+
+    public static <T, R> R executeWithRetry(Supplier<T> inputSupplier, Function<T, R> processFunction) {
+        while (true) {
+            try {
```
들여쓰기가 두단계네요!

Ryan-Dia replied:
해당 부분 수정했습니다.
감사합니다.

Ryan-Dia committed:
style: 변수명 수정

ENTER -> LINE_SEPARATOR

Ryan-Dia committed:
style(LottoController): 변수명 수정

generateLotto -> processLottoGeneration

Ryan-Dia committed:
style: 매직넘버 상수화

보너스 볼이 일치하는지 비교하는 기준점인 숫자 5를 상수화

Ryan-Dia committed:
style: 코드 가독성을 위해 스트림 체이닝 개행

Ryan-Dia committed:
refactor(RecoveryUtils): 인덴트 1로 맞추기위한 메서드 분리

Ryan-Dia committed:
refactor: toLottoNumberList 메서드 이동 및 분리

- 기존 Lotto에 있던 toLottoNumberList 메서드를 분리
- 검증과 리스트로 만드는 일을 분리
- parseLottoNumbers 추가
- validateInputWinnerNumbers 추가

Ryan-Dia committed:
refactor: rankCounts 필드에 final, private 키워드 추가

Ryan-Dia committed:
refactor: getMatchCount 메서드 개선

Lotto에서 로또 넘버 list를 get하지 않고 contains메서드를 만들어서 비교하도록 변경

Ryan-Dia committed:
refactor: Lotto 클래스에서 정적 팩터리 메서드 제거

Ryan-Dia committed:
style: 메서드명 변경(generate -> processLottoTicketGeneration)

Ryan-Dia committed:
refactor: 최대 재시도 횟수를 추가하여 무한시도를 할 수 없도록 변경

Ryan-Dia commented:
안녕하세요 샐리 리뷰해주셔서 감사합니다.  ☺️

질문 사항에도 답변해주셔서 감사합니다.
많은 도움이 되었습니다.

피드백주신 부분은 전부 수정하였습니다. 

더불어 조금 개선할 점들이 있어서 해당 부분도 수정하였습니다.

## ⚒️ 개선한 점

### 1. 변수명/메서드명 수정
- ENTER -> LINE_SEPARATOR
- 5 ->  BONUS_CHECK_POINT
- generateLotto -> processLottoGeneration
- generate -> processLottoTicketGeneration

### 2. 가독성 향상
- 스트림 체이닝 개행

### 3. 인덴트 1로 맞추기 위해 메서드 분리

### 4. toLottoNumberList 메서드 분리
- 입력값에 대한 유효성과 객체 생성을 분리

### 5. Lotto에서 로또 넘버 list를 get하지 않고 contains메서드를 만들어서 비교하도록 변경

### 6.  Lotto 클래스에서 정적 팩터리 메서드 제거

### 7. 키워드 추가 
- Map<Rank, Integer> rankCounts -> private final Map<Rank, Integer> rankCount;

### 8. 최대 재시도 횟수 제한
최대 재시도 횟수 5회를 제한하여 무한시도를 할 수 없도록 변경

Ryan-Dia committed:
test: Moeny 도메인 테스트 코드 추가

Ryan-Dia committed:
refactor: 불필요한 정적 팩터리 메서드 제거

Ryan-Dia committed:
refactor: validateAmount 메서드 키워드 변경 (public->private)

Ryan-Dia committed:
test: 로또 도메인 테스트 추가

Ryan-Dia committed:
refactor: 오름차순으로 정렬이 되어있지 않으면 예외가아니라 보정하도록 수정

dusdn1702 commented:
안녕하세요 새로이!
피드백 반영해주신 부분들 확인하였습니다 ㅎㅎ
사실, 새로이처럼 명확한 이유가 있다면 저는 같이 생각해보기 위한 질문만 던져드리는 것이기 때문에 모두 반영해주시지 않아도 괜찮아요 😄 
다시 한번 생각해보고, 함께 이야기하면서 많은 것을 얻어가는 우테코 생활이 되셨으면 좋겠습니다 👍 
짧은 시간에 테스트도 추가해주시고 글도 작성해주셔서 감명 깊었어요 💯 
다만, 이번엔 두 도메인에 대한 단위 테스트만 작성해주셨는데 다음 단계에서는 온보딩 기간에 배웠던 테스트들을 더 많이 적용해보는 건 어떨까요? 
첫 미션 첫 단계 고생 많으셨습니다 ! 다음 단계도 화이팅 입니다 💪

dusdn1702 reviewed (approved):
안녕하세요 새로이!
피드백 반영해주신 부분들 확인하였습니다 ㅎㅎ
사실, 새로이처럼 명확한 이유가 있다면 저는 같이 생각해보기 위한 질문만 던져드리는 것이기 때문에 모두 반영해주시지 않아도 괜찮아요 😄
다시 한번 생각해보고, 함께 이야기하면서 많은 것을 얻어가는 우테코 생활이 되셨으면 좋겠습니다 👍
짧은 시간에 테스트도 추가해주시고 글도 작성해주셔서 감명 깊었어요 💯
다만, 이번엔 두 도메인에 대한 단위 테스트만 작성해주셨는데 다음 단계에서는 온보딩 기간에 배웠던 테스트들을 더 많이 적용해보는 건 어떨까요?
첫 미션 첫 단계 고생 많으셨습니다 ! 다음 단계도 화이팅 입니다 💪
